qiskit/pulse/transforms/alignments.py

```diff
@@ -0,0 +1,454 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""A collection of passes to reallocate the timeslots of instructions according to context."""
+
+import abc
+from typing import Optional, Iterable, Callable, Dict, Any
+
+import numpy as np
+
+from qiskit.pulse import channels as chans, instructions
+from qiskit.pulse.exceptions import PulseError
+from qiskit.pulse.schedule import Schedule, ScheduleComponent
+
+
+class AlignmentTransform(abc.ABC):
+    """An abstract class for schedule alignment."""
+    is_sequential = None
+
+    @abc.abstractmethod
+    def align(self, schedule: Schedule) -> Schedule:
+        """Reallocate instructions according to the policy.
+
+        Only top-level sub-schedules are aligned. If sub-schedules are nested,
+        nested schedules are not recursively aligned.
+
+        Args:
+            schedule: Schedule to align.
+
+        Returns:
+            Schedule with reallocated instructions.
+        """
+        pass
+
+    @property
+    def context(self) -> Dict[str, Any]:
+        """Returns dictionary to represent this alignment."""
+        return {'alignment': self.__class__.__name__}
+
+    def __eq__(self, other):
+        """Check equality of two transforms."""
+        return isinstance(other, type(self)) and self.context == other.context
+
+
+class AlignLeft(AlignmentTransform):
+    """Align instructions in as-soon-as-possible manner.
+
+    Instructions are placed at earliest available timeslots.
+    """
+    is_sequential = False
+
+    def align(self, schedule: Schedule) -> Schedule:
+        """Reallocate instructions according to the policy.
+
+        Only top-level sub-schedules are aligned. If sub-schedules are nested,
+        nested schedules are not recursively aligned.
+
+        Args:
+            schedule: Schedule to align.
+
+        Returns:
+            Schedule with reallocated instructions.
+        """
+        aligned = Schedule()
+        for _, child in schedule._children:
+            _push_left_append(aligned, child)
```

Comment:
```suggestion
            _push_left_append(aligned, child)
```

It may be nice to make this a class method now.